internal/worker_versioning_rules.go

```diff
@@ -0,0 +1,338 @@
+		TargetBuildId string
+		Force         bool
+	}
+)
+
+func (uw *UpdateWorkerVersioningRulesOptions) validateAndConvertToProto(namespace string) (*workflowservice.UpdateWorkerVersioningRulesRequest, error) {
+	if namespace == "" {
+		return nil, errors.New("missing namespace argument")
+	}
+	if uw.TaskQueue == "" {
+		return nil, errors.New("missing TaskQueue field")
+	}
+	if !uw.Operation.isValidUpdateVersioningOp() {
+		return nil, errors.New("invalid Operation BuildID field")
+	}
+	req := &workflowservice.UpdateWorkerVersioningRulesRequest{
+		Namespace:     namespace,
+		TaskQueue:     uw.TaskQueue,
+		ConflictToken: uw.ConflictToken.token,
+	}
+
+	switch v := uw.Operation.(type) {
+	case *VersioningOpInsertAssignmentRule:
+		req.Operation = &workflowservice.UpdateWorkerVersioningRulesRequest_InsertAssignmentRule{
+			InsertAssignmentRule: &workflowservice.UpdateWorkerVersioningRulesRequest_InsertBuildIdAssignmentRule{
+				RuleIndex: v.RuleIndex,
+				Rule:      versioningAssignmentRuleToProto(&v.Rule),
+			},
+		}
+	case *VersioningOpReplaceAssignmentRule:
+		req.Operation = &workflowservice.UpdateWorkerVersioningRulesRequest_ReplaceAssignmentRule{
+			ReplaceAssignmentRule: &workflowservice.UpdateWorkerVersioningRulesRequest_ReplaceBuildIdAssignmentRule{
+				RuleIndex: v.RuleIndex,
+				Rule:      versioningAssignmentRuleToProto(&v.Rule),
+				Force:     v.Force,
+			},
+		}
+	case *VersioningOpDeleteAssignmentRule:
+		req.Operation = &workflowservice.UpdateWorkerVersioningRulesRequest_DeleteAssignmentRule{
+			DeleteAssignmentRule: &workflowservice.UpdateWorkerVersioningRulesRequest_DeleteBuildIdAssignmentRule{
+				RuleIndex: v.RuleIndex,
+				Force:     v.Force,
+			},
+		}
+	case *VersioningOpInsertRedirectRule:
+		req.Operation = &workflowservice.UpdateWorkerVersioningRulesRequest_InsertCompatibleRedirectRule{
+			InsertCompatibleRedirectRule: &workflowservice.UpdateWorkerVersioningRulesRequest_AddCompatibleBuildIdRedirectRule{
+				Rule: versioningRedirectRuleToProto(&v.Rule),
+			},
+		}
+	case *VersioningOpReplaceRedirectRule:
+		req.Operation = &workflowservice.UpdateWorkerVersioningRulesRequest_ReplaceCompatibleRedirectRule{
+			ReplaceCompatibleRedirectRule: &workflowservice.UpdateWorkerVersioningRulesRequest_ReplaceCompatibleBuildIdRedirectRule{
+				Rule: versioningRedirectRuleToProto(&v.Rule),
+			},
+		}
+	case *VersioningOpDeleteRedirectRule:
+		req.Operation = &workflowservice.UpdateWorkerVersioningRulesRequest_DeleteCompatibleRedirectRule{
+			DeleteCompatibleRedirectRule: &workflowservice.UpdateWorkerVersioningRulesRequest_DeleteCompatibleBuildIdRedirectRule{
+				SourceBuildId: v.SourceBuildId,
+			},
+		}
+	case *VersioningOpCommitBuildId:
+		req.Operation = &workflowservice.UpdateWorkerVersioningRulesRequest_CommitBuildId_{
+			CommitBuildId: &workflowservice.UpdateWorkerVersioningRulesRequest_CommitBuildId{
+				TargetBuildId: v.TargetBuildId,
+				Force:         v.Force,
+			},
+		}
+	}
```

Comment:
IMO it's best to add a `default:` clause that returns an error for unrecognized type

Reply:
Yes, I keep forgetting this is not Rust...